### ISSUE TITLE
EHPI generation

### ISSUE THREAD

**Issue by vinicius121** [noboevbo/ehpi_action_recognition#2]
Hi,

Very interesting paper.
I was looking at your code to generate the EHPI images and I have some questions.
```
for human_id, action_vecs in humans_for_action_rec.items():
			ehpi_img = np.zeros((32, 15, 3), dtype=np.float32)
			for frame_num, action_vec in enumerate(action_vecs):
				if action_vec is None:
					continue
				ehpi_img[frame_num] = action_vec
			ehpi_img = np.transpose(ehpi_img, (2, 0, 1))
			# Set Blue Channel to zero
			ehpi_img[2, :, :] = 0
			# Normalize EHPI
			tmp_dict = {'x': ehpi_img}
			tmp_dict['x'] = self.remove(tmp_dict)['x']
			ehpi_img = self.normalize(tmp_dict)['x']
			net_input = np.zeros((1, 3, 32, 15), dtype=np.float32)
			net_input[0] = ehpi_img

			input_seq = Variable(torch.tensor(net_input, dtype=torch.float)).cuda()
			tag_scores = self.model(input_seq).data.cpu().numpy()[0]
			outputs[human_id] = tag_scores
		return outputs
```
How would I change the code to encode x, y, an z information and normalize it?
I'm having some issues in understanding how did you normalize each coordinate.
Could you provide a script to generate  and visualize the EHPI images?

Thanks

**Comment by denclade** [comment]
Hi, 
sorry for the late response. The basic process to generate an EHPI is best displayed in the dataset loader (https://github.com/noboevbo/nobos_torch_lib/blob/master/nobos_torch_lib/datasets/action_recognition_datasets/ehpi_dataset.py). It loads data in shape [ SequenceNum, Channel, Width, Height] which contains raw 2D Joint Positions (X, Y) as well as the Score of the joint in Channel 3. If you would like to encode a z information you could set it to X_[:, 2, :, :] = z (See load_X in the dataloader). Also check the __get_item__ method which applies the required transformations to the 2D coordinates. As you can see in the code example you posted the remove and normalization are applied in inference mode. You would have to extend the RemoveJointsEhpi() and NormalizeEhpi() to include the z coordinate, currently it supports only x and y coordinates (adding [2, :, :] = bla). 

To visualize a ehpi simply transpose the EHPI to img shape, multiply the normalized values by 255 and change the datatype to uint8. Like follows:
```
action_img = np.transpose(ehpi_img, (2, 1, 0))
action_img *= 255
action_img = action_img.astype(np.uint8)
# action_img = cv2.resize(action_img, (action_img.shape[1] * 30, action_img.shape[0] * 30), cv2.INTER_NEAREST)
action_img = cv2.cvtColor(action_img, cv2.COLOR_BGR2RGB)
cv2.imshow("ehpi", action_img)
cv2.waitKey(1)
```
The resize is just for visualization purpose, but I had some problems with the nearest neighbour interpolation in cv2 resulting in a "blurred" EHPI, thus I scaled the images up in post processing.
I hope this helps you out a bit.

**Comment by vinicius121** [comment]
Hi,

Thanks for your reply.
What are the transformations exactly that you apply to the 2D coordinates in **get_item**?
The dataset that i have has the world 3D coordinates. Do you think that i have to do some transformation?

Thanks

**Comment by denclade** [comment]
As you can see in the training and inference code we apply a normalization of the x and y values to a range from 0 to 1 based on the min and max values over the 32 frame EHPI window. We also remove joints outside the image dimension. For training we apply DataAugmentation Transforms (Remove Joints, Flip and Scale).
If you only work with world 3D Coordinates in Train and Test and no camera sensors etc. are involved you usually don't need any special transformations. Most important is to extend NormalizeEhpi to support the z value, accordingly to the x and y values (setting z_min to 0, get z_max, normalize 0-zmax to 0-1). For training I would also add z value support to the RemoveJointsEhpi, FlipEhpi and ScaleEhpi.

**Comment by vinicius121** [comment]
Hi,

I'm also doing data augmentation (Remove joints, Flip and scale).  And I also added the support for the z coordinate in my case for the NormalizeEhpi.
Concerning the data, it was recorded using a Kinect sensor. And the idea is to use the trained model and the Kinect sensor to predict the trained actions. Since Kinect already returns x,y,z in world coordinates i was thinking on not doing any special transformations. What do you think in your experience?

Thanks

**Comment by denclade** [comment]
If you added support for the z value and remove the = 0 assignments currently in the code it should work well with Kinect 3D coordinates.
